### ISSUE TITLE
Write file in browser

### ISSUE THREAD

**Issue by xSirrioNx** [guyonroche/exceljs#402]
Hi!
I don't understand how to save created workbook to file in browser =(
Can anybody explain it to me?

**Comment by barbalex** [comment]
did you check this: https://github.com/guyonroche/exceljs#browser

**Comment by GenuineRex** [comment]
`        var wb = new ExcelJS.Workbook();
        var ws = wb.addWorksheet('Export');

        ws.getCell('A1').value = 7;
        ws.getCell('B1').value = 'Hello, World!';

        wb.xlsx.writeBuffer( {
                base64: true
            })
            .then(function (xls64) {
                console.log('Done.');

                // build anchor tag and attach file (works in chrome)
                //
                var a = document.createElement("a");
                var data = new Blob([xls64], { type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" });
                var url = URL.createObjectURL(data);

                a.href = url;
                a.download = "export.xlsx";

                document.body.appendChild(a);

                a.click();
                setTimeout(function() {
                        document.body.removeChild(a);
                        window.URL.revokeObjectURL(url);
                    },
                    0);

            })
            .catch(function(error) {
                console.log(error.message);
            });`

**Comment by jathanasiou** [comment]
For the record, the above only seems to work for `xlsx`. The `wb.csv` object has no `writeBuffer()` method implemented in its prototype.

**Comment by GenuineRex** [comment]
I have not tested csv but a quick look at the source shows a writeBuffer method.

**Comment by jathanasiou** [comment]
I noticed it too. I was added to `master` a few days ago and hasn't been released yet though. Hopefully it will fully resolve this issue.

With the current `1.2.1` release, csv writing doesn't work in a browser via Buffer.

**Comment by GenuineRex** [comment]
Using csv.write would work just as well but you would need to add your own stream (which is all that writeBuffer does for you).

**Comment by jathanasiou** [comment]
Agreed. I tried doing that a little pre-`1.0.0` and seem to recall finding it non-trivial to implement. Regardless, the `writeBuffer` method needs to work on csv too the sake of completeness and then this issue could be closed.

**Comment by meiyoufengzhengdexian** [comment]
When it generates more than 40,000 rows of excel on the browser side, it crashes.

**Comment by graciano** [comment]
> 
> 
> When it generates more than 40,000 rows of excel on the browser side, it crashes.

this seems to be a memory issue, not from this library

**Comment by kenberkeley** [comment]
Based on [GenuineRex's solution](https://github.com/exceljs/exceljs/issues/402#issuecomment-347927095) and https://github.com/SheetJS/sheetjs#writing-workbooks :

```
$ yarn add exceljs file-saver
```

```js
import Excel from 'exceljs'
import { saveAs } from 'file-saver'

const workbook = new Excel.Workbook()

const worksheet = workbook.addWorksheet('People')

worksheet.columns = [
  { header: 'Id', key: 'id', width: 10 },
  { header: 'Name', key: 'name', width: 10 },
  { header: 'D.O.B.', key: 'DOB', width: 30 }
]

worksheet.addRow([10086, 'Ken1', 'YYYY-MM-DD'])
worksheet.addRow([10087, 'Ken2', 'YYYY-MM-DD'])
worksheet.addRow([10088, 'Ken3', 'YYYY-MM-DD'])

;(async () => {
  const xls64 = await workbook.xlsx.writeBuffer({ base64: true })
  const fileName = 'xxx'
  saveAs(
    new Blob([xls64], { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' }),
    fileName
  )
})()
```
